### PR TITLE
Extend Inductor to support the third-party backend

### LINKED ISSUE #99419: Extend TorchInductor to support more backends
### 🚀 The feature, motivation and pitch

This RFC proposes supporting TorchInductor to add a new backend at runtime to generate code for a particular device.

### Motivation
Currently, there are two integration levels to add a new backend for PT compiler: at the AtenIR/PrimsIR level as a new backend of Dynamo, and at the Inductor loop IR level as a new codegen backend of the Inductor.

- Integrating a new backend at the AtenIR/PrimsIR level is straightforward and provides more freedom to the new backend to optimize the captured graph. But it might be suboptimal performance if the backend lacks the DL compiler capability because the decomposed operations for a single operation might need more memory and worse data locality compared with the single operation if the decomposed operations can't be fused.

- Integrating at the Inductor loop IR level can significantly simplify the complexity of design and implementation by leveraging the Inductor's fusion capability and other optimizations directly. The new backend just needs to focus on how to generate optimal code for a particular device.

This RFC focuses on the latter one.

### Proposal
There are currently two backends for the Inductor - C++/OpenMP and Triton. And we propose providing a dynamic registration mechanism on the Inductor side for a new backend. The new backend can be out-of-tree and register its codegen for a particular device at runtime.

It can provide more flexibility as a new backend just needs to customize three fundamental structures: `Scheduling`, `Kernel` and `WrapperCodegen`. The sample code could be as follows:

```python

# in-tree/out-of-tree
class ExtensionKernel(Kernel):
    pass

# in-tree/out-of-tree
class ExtensionScheduling(Scheduling):
    ...
    def codegen_nodes(self, nodes):
        with ExtensionKernel(...):
            pass
    ...

# in-tree/out-of-tree
class ExtensionWrapperCodeGen(WrapperCodeGen):
    ...

```
The `ExtensionKernel` class and `ExtensionScheduling` class are used to define the operations and scheduling, respectively, for the new backend. The `ExtensionWrapperCodeGen` class generates the Python wrapper code for the backend to glue the kernel. All three classes can be out-of-tree as they can be registered to the Inductor at runtime.

Besides the key structures, we will provide two utility functions to support the runtime registration - `register_scheduling_for_device` and `get_scheduling_for_device`. The two utility functions are used to register and get the backend for a particular device, respectively.
```python
# in-tree
def register_scheduling_for_device(device:str, ext_scheduling: Scheduling):
    pass

# in-tree
def get_scheduling_for_device(device:str) -> Scheduling:
    pass
```

After that, we just need to add a default path to route to the newly added backend according to the device type.
```python
# in-tree
def create_backend(self, device: torch.device):
    ...
    if device.type == "cpu":
        from .codegen.cpp import CppScheduling
        return CppScheduling(self)
    elif device.type == "cuda":
        from .codegen.triton import TritonScheduling
        return TritonScheduling(self)
    else:
        return get_scheduling_for_device(device.type)(self)
```

### Alternatives

Another option is to add a new backend-related code to Inductor directly. Take the backend creation as an example, the code sample could be as follows:

```python
# in-tree
class NewBackendKernel(Kernel):
    pass

# in-tree
class NewBackendScheduling(Scheduling):
    ...
    def codegen_nodes(self, nodes):
        with NewBackendKernel(...):
            pass
    ...

# in-tree
class NewBackendWrapperCodeGen(WrapperCodeGen):
    ...

# in-tree
def create_backend(self, device: torch.device):
    ...
    if device.type == "cpu":
        from .codegen.cpp import CppScheduling
        return CppScheduling(self)
    elif device.type == "xpu":
        from .codegen.new_device import NewBackendScheduling
        return NewBackendScheduling(self)
    else:
        from .codegen.triton import TritonScheduling
        return TritonScheduling(self)
```

Compared to the former option, the latter is more straightforward and simple. But it requires all the new backend-related code to be in-tree. In other words, the new backend must be landed the PyTorch. It is a high bar to support other devices.

Currently, the common practice to support new devices is through PyTorch Extension(out of the pytorch code tree), where hardware vendors implement all the necessary operations in their own PyTorch extensions. And it might take longer to finalize the design, maturize the feature, and some other things to meet the bar of upstreaming. Before it is a part of PyTorch, it will conflict with this alternative's "in-tree" requirement.

Besides that, it's worth noting that the CI/CD pipeline for the latter option may be more challenging since new devices are typically not available on AWS.

### Additional context

_No response_

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100706

This PR intends to extend Inductor to support the third-party backend that only focuses on the code generation just like what C++/OpenMP and Triton backend have done. 

Currently, the generated code by Inductor contains two major parts. One is the kernel, and the other is the Python wrapper to glue the kernel. Therefore, the third-party backend needs to customize the two parts to generate its specific code.

- Python wrapper code generation

  Inductor provides a `WrapperCodeGen` class to generate the Python wrapper code to glue the kernel. Therefore, it is straightforward for the third-party backend to generate the backend-specific Python wrapper code. It just needs to inherit the `WrapperCodeGen` class and purposely override the particular member functions. 

- Kernel code generation
  
  It is driven by different `Scheduling`. Hence, the third-party backend needs to provide a custom `Scheduling` for its specific kernel code generation. Currently, `CppScheduling` and `TritonScheduling` are for C++/OpenMP and Triton backend, respectively. But there is no common `Scheduling` class. Based on the scheduling invocation, this PR abstracts a common `Scheduling` class containing the following member functions. 

  -   [group_fn](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL649C64-L649C64)
  - [flush](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1150)
  - [can_fuse_vertical](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1006)
  - [can_fuse_horizontal](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1008C45-L1008C64)
  - [codegen_template](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234) _This function is only available for triton. If the third-party backend behaves as a sub-class of `TritonScheduling`, it can override it or reuse it._
  - [codegen_nodes](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#L1234)
  - [codegen_sync](https://github.com/pytorch/pytorch/blob/71c4becda7cbdea8590cb98da9f2ea4c1edc0cc1/torch/_inductor/scheduler.py#LL1251C1-L1251C1). _This function is only available for triton debug purpose. But it might also be useful for other computation devices. Therefore, we'd prefer to keep this function._ 
  
  The third-party backend needs to inherit from the `Scheduling` class and implement these functions.

Regarding some other classes like `CppKernel` and `TritonKernel` for code generation, they are used by or part of the logic of either `Scheduling` or `WrapperCodeGen`. Hence, this PR does not define the interface and leaves the flexibility to the third-party backend. The third-party backend can decide to implement these classes from scratch or reuse them by inheriting and overriding them.

cc @voznesenskym @penguinwu @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @soumith @desertfire

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -633,7 +634,18 @@ def init_wrapper_code(self):
                 )
                 return
 
-        self.wrapper_code = WrapperCodeGen()
+        if not all(device_type in ["cuda", "cpu"] for device_type in self.device_types):
+            device_types_copy = self.device_types
+            device_types_copy.discard("cuda")
+            device_types_copy.discard("cpu")
```
Comment (jgong5) [inline torch/_inductor/graph.py:640]: this changes `self.device_types` inplace?

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -1075,3 +1078,18 @@ def triton_config_to_hashable(cfg):
     items.append(("num_warps", cfg.num_warps))
     items.append(("num_stages", cfg.num_stages))
     return tuple(items)
+
+
+def register_backend_for_device(
```
Comment (jgong5) [inline torch/_inductor/utils.py:1083]: better to make it a decorator on custom `Scheduling` and `WrapperCodegen"?

File: torch/_inductor/codegen/scheduling.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    @abstractmethod
+    def group_fn(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def codegen_template(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def codegen_nodes(self, *args, **kwargs):
```
Comment (jgong5) [inline torch/_inductor/codegen/scheduling.py:22]: why it is not a list of scheduling nodes?

File: torch/_inductor/codegen/scheduling.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    @abstractmethod
+    def can_fuse_horizontal(self, node1, node2):
+        pass
+
+    @abstractmethod
+    def group_fn(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def codegen_template(self, *args, **kwargs):
```
Comment (jgong5) [inline torch/_inductor/codegen/scheduling.py:18]: why is it not template node and epilogues?

File: torch/_inductor/codegen/scheduling.py
Comment (jgong5) [inline torch/_inductor/codegen/scheduling.py:1]: better to type the methods in `Scheduling` and avoid using `*args` and `**kwargs` if possible.

File: torch/_inductor/graph.py [hunk 5]
```diff
@@ -633,7 +634,18 @@ def init_wrapper_code(self):
...
+            assert len(device_types_copy) == 1, "Does not support mixing {}".format(
+                "+".join(device_types_copy)
+            )
+            wrapper_code_gen_cls = get_wrapper_for_device(device_types_copy.pop())
+            assert wrapper_code_gen_cls
```
Comment (jgong5) [inline torch/_inductor/graph.py:645]: add error message here.

File: torch/_inductor/utils.py [hunk 6]
```diff
@@ -31,6 +31,9 @@
 
 VarRanges = Dict[sympy.Expr, sympy.Expr]
 
+extension_scheduling: Dict[str, type] = {}
```
Comment (jansel) [inline torch/_inductor/utils.py:34]: utils.py seems like the wrong place for this registration stuff.  Either a new file or scheduler.py.

File: torch/_inductor/scheduler.py [hunk 7]
```diff
@@ -1184,6 +1190,8 @@ def create_backend(self, device: torch.device):
             from .codegen.triton import TritonScheduling
 
             return TritonScheduling(self)
+        elif get_scheduling_for_device(device.type):
```
Comment (jansel) [inline torch/_inductor/scheduler.py:1193]: We should refactor the lines above to use the same registration system.

File: torch/_inductor/codegen/scheduling.py (hunk 4 above)
Comment (jansel) [reply torch/_inductor/codegen/scheduling.py:18]: I don't think this will work with non-triton backends currently.

File: torch/_inductor/codegen/scheduling.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        pass
+
+    @abstractmethod
+    def codegen_nodes(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def codegen_sync(self, *args, **kwargs):
```
Comment (jansel) [inline torch/_inductor/codegen/scheduling.py:26]: Use actual function signatures not args/kwargs.

File: torch/_inductor/codegen/scheduling.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+from abc import ABC, abstractmethod
+
+
+class Scheduling(ABC):
```
Comment (jansel) [inline torch/_inductor/codegen/scheduling.py:4]: We should modify the existing backends to inherit from this. 

Add docstrings for every method.

Perhaps this belongs in common.py, or some sort of backend registration file.

File: test/inductor/extension_backends/extension_scheduling.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    def codegen_nodes(self, *args, **kwargs):
+        nodes = args[0]
+        self._scheduling.codegen_nodes(nodes)
+
+    def codegen_sync(self, *args, **kwargs):
+        pass
+
+    def flush(self, *args, **kwargs):
+        self._scheduling.flush()
```
Comment (jansel) [inline test/inductor/extension_backends/extension_scheduling.py:37]: I was hoping to see a more complete example than this.  One that actually generates some minimal working C++ code.  I worry there are hidden issues that will be uncovered.

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/inductor/extension_backends/extension_scheduling.py (hunk 10 above)
Comment (EikanWang) [reply test/inductor/extension_backends/extension_scheduling.py:37]: Do you think partial elementwise operations are good enough? Or should we implement all the functions?

Comment (jansel) [reply test/inductor/extension_backends/extension_scheduling.py]: It doesn't need to implement everything, but it should demonstrate the key concepts:
1) pointwise ops
2) reduction ops
3) horizontal fusion
4) vertical fusion

**Comment by jgong5** [comment]
BTW, can we also support overriding the backend of registered devices?

File: test/inductor/extension_backends/extension_scheduling.py (hunk 10 above)
Comment (EikanWang) [reply test/inductor/extension_backends/extension_scheduling.py]: Okay. I will develop a minimal codegen to demonstrate these key concepts.

**Comment by EikanWang** [comment]
> BTW, can we also support overriding the backend of registered devices?

In practice, this PR should have supported this scenario.
https://github.com/pytorch/pytorch/pull/100706/files#diff-e7f291bf7d847fea77a46bd4c0ce6ee0eacf851714a397b4271e54f36fedf369R672

**Comment by EikanWang** [comment]
Just update the status here. We are working on a custom Trtion backend for Intel GPU.

File: torch/_inductor/codegen/common.py [hunk 11]
```diff
@@ -38,6 +38,46 @@ def data_type_logger(msg):
...
+#
+# Other classes, such as CppKernel and TritonKernel, used for code generation, typically form part
+# of the logic for either Scheduling or WrapperCodeGen. So the Scheduling and WrapperCodeGen interfaces
+# provide flexibility to the backend. A backend can choose to implement these classes from scratch,
+# or reuse them by extending and overriding as necessary. And Inductor provides the registration API,
+# register_backend_for_device, to equip a new backend at runtime.
+#
+# Intel has developed a new backend on top of Triton to support Intel GPUs, leveraging these interfaces.
+# This backend can be used as a reference:
+# https://github.com/intel/intel-extension-for-pytorch/blob/5dcc9d57e5422cf295e1a1ee97896d6b6a554a85/intel_extension_for_pytorch/_inductor/__init__.py#L9
```
Comment (EikanWang) [inline torch/_inductor/codegen/common.py:65]: @jansel , I added some comments here and provided the example code of IPEX.

**Review by jansel — APPROVED** [review]

**Comment by yanboliang** [comment]
@EikanWang This PR introduced regression of compile time for torchbench and huggingface models. I sent #106652 to revert it. Can you investigate issue and re-submit this?

The steps of benchmark compilation time:
* Run ```python benchmarks/dynamo/torchbench.py -dcuda --training --performance --inductor```
* Check the generated ```inductor.csv``` file.
* There is a column ```compilation_latency```.

Or you can run single model benchmark:
* Run ```python benchmarks/dynamo/torchbench.py -dcuda --training --performance --inductor --only hf_T5```

The following results are from my benchmarks:
```
dev,name,batch_size,speedup,abs_latency,compilation_latency,compression_ratio,eager_peak_mem,dynamo_peak_mem,calls_captured,unique_graphs,graph_breaks,unique_graph_breaks
cuda,hf_T5,8,1.229514,133.804266,48.471589,1.016064,19.291709,18.986706,795,1,8,7
cuda,hf_T5,8,1.227059,133.983976,47.352921,1.016064,19.291709,18.986706,795,1,8,7
cuda,hf_T5,8,1.234702,133.383399,54.385957,1.016064,19.291709,18.986706,795,1,8,7
cuda,hf_T5,8,1.226521,134.050623,56.280598,1.016064,19.291709,18.986706,795,1,8,7
```

The first two rows: not including this PR.
The last two rows: including this PR.

**Comment by EikanWang** [comment]
@yanboliang , I will check the compilation latency and resubmit this PR.

**Comment by EikanWang** [comment]
Regarding the C++/OpenMP backend,  `codecache.pick_vec_isa()` to check vectorization ISA is a time-consuming and one-shot operation. It leads to taking a longer time to import `codegen.cpp` package because the `LoopLevel` of the package is decorated by `@dataclasses.dataclass` while the decorator will invoke `codecache.pick_vec_isa()` to initialize the `simd_nelements` of the `LoopLevel`.
https://github.com/pytorch/pytorch/blob/c14cf312c9302be2cafd3be61484f5f9613fab63/torch/_inductor/codegen/cpp.py#L2883C53-L2883C53

In terms of the Triton backend, it does not need to touch it. But we'd prefer to uniform the code. Therefore, the new design registers `CpuScheduling` for CPU and `TritonScheduling` for Triton simultaneously regardless of whether the current backend is Triton. It will bring additional overhead to the Triton backend. 

```python
def init_backend_registration(self):
    if get_scheduling_for_device("cpu") is None:
        from .codegen.cpp import CppScheduling

        register_backend_for_device("cpu", CppScheduling, WrapperCodeGen)

    if get_scheduling_for_device("cuda") is None:
        from .codegen.triton import TritonScheduling

        register_backend_for_device("cuda", TritonScheduling, WrapperCodeGen)
```

To resolve the compilation latency regression for the Triton backend, we changed the `LoopLevel` a little bit by moving the `simd_nelements` to `__post_init__` and the compilation performance would be back. The new PR is https://github.com/pytorch/pytorch/pull/106874

**Comment by EikanWang** [comment]
@yanboliang , I resubmitted this PR #106874
